lib/index.js

```diff
@@ -2108,6 +2108,188 @@ addMethod('toPromise', function (PromiseCtor) {
+ * @private
+ *
+ * createObserver(
+ *     function (x) { console.log(x); },
+ *     function (err) { console.error(err); },
+ *     function () { console.log('done'); }
+ * )
+ *
+ * createObserver(
+ *     null,
+ *     null,
+ *     function () { console.log('done'); }
+ * )
+ *
+ * createObserver({
+ *     next: function (x) { console.log(x); },
+ *     error: function (err) { console.error(err); },
+ *     complete: function () { console.log('done'); }
+ * })
+ */
+function createObserver (onNext, onError, onComplete) {
+    var isObserver = onNext && !onNext.call && typeof onNext === 'object';
+
+    // ensure if we have an observer that we don't also have callbacks. Users
+    // must choose one.
+    if (isObserver && (onError || onComplete)) {
+        throw new Error('Subscribe requires either an observer object or optional callbacks.');
+    }
+
+    // onNext is actually an observer
+    if (isObserver) {
+        return onNext;
+    }
+
+    // Otherwise create an observer object
+    var observer = {};
+
+    if (_.isFunction(onNext)) {
+        observer.next = onNext;
+    }
+
+    if (_.isFunction(onError)) {
+        observer.error = onError;
+    }
+
+    if (_.isFunction(onComplete)) {
+        observer.complete = onComplete;
+    }
+
+    return observer;
+}
+
+/**
+ * Consumes values using the Observable subscribe signature.
+ *
+ * Implements the Observable subscribe functionality as defined by the spec:
+ * https://tc39.github.io/proposal-observable/#observable-prototype-subscribe
+ *
+ * @id toPromise
+ * @section Consumption
+ * @name Stream.subscribe(onNext, onError, onComplete)
+ * @param {Function|object|null} onNext - Handler for next value or observer
+ * @param {Function|null} onError - Handler function for errors.
+ * @param {Function|null} onCompleted - Handler Function when stream is done.
+ * @api public
+ *
+ *  // with callbacks
+ * _([1, 2, 3, 4]).subscribe(
+ *     function onNext (x) {
+ *         // Called for each value that comes downstream
+ *         console.log('Received onNext value', x);
+ *     },
+ *     function onError (err) {
+ *         // Called one time with error or zero if no errors occur upstream
+ *         console.error('Single highland stream error', err);
+ *     },
+ *     function onComplete () {
+ *         // Receives no arguments
+ *         // Called only once when stream is completed.
+ *         console.log('Completed!');
+ *     }
+ * );
+ *
+ * // with an observer
+ * _[1, 2, 3, 4]).subscribe({
+ *     next (x) {
+ *         console.log('Received next value', x);
+ *     },
+ *     error (err) {
+ *         console.error('An error occurred upstream', err);
+ *     },
+ *     complete () {
+ *         console.log('Completed!')
+ *     }
+ * });
+ */
+
+addMethod('subscribe', function (onNext, onError, onComplete) {
+    var self = this;
+
+    // Don't let users subscribe to an already completed stream
+    if (self._nil_pushed) {
```

Comment:
Check `self.done` instead of `self._nil_pushed`. Just because `nil` has been pushed doesn't mean it's been emitted. It's possible that the source ignored backpresure and just pushed everything onto the stream buffer at once.

Also. It's debatable, but I think we should call `observer.error(new Error('...')` instead of throwing an error here. The rationale is that people don't usually expect methods like `subscribe` to throw. The user is probably more likely to notice an error that's delivered through the observer.

Reply:
Thanks for the clarification on self.done. As for the error, not everyone is going to provide an error handler which was why I threw it. Maybe we should just call console.error directly as more of a warning instead of throwing? Alternatively we could throw if no error handler is provided but that conditional behavior may not be desired.

Reply:
Having some trouble with done as it’s always a truthy function, do you by chance mean `self.ended`?

Reply:
Using self.ended seems to work, even highlighted a nuance my original solution would have missed:

If you did:
```js
var stream = _.empty();
stream.subscribe();
```

It would have thrown an error. Now it only throws an error if it's been consumed in some way which I understand to be the behavior you expect. Anyway, added another test case to mark that difference.

Reply:
`self.ended` is correct. I mis-remembered which one was which.

> As for the error, not everyone is going to provide an error handler which was why I threw it. Maybe we should just call console.error directly as more of a warning instead of throwing? Alternatively, we could throw if no error handler is provided and use observer.error when it is but that conditional behavior may not be desired.

The conditional behavior isn't always bad---promises have unhandled promise exceptions which are basically the same thing. However, observables don't have the same mechanism, so I'd stray away from inventing one that only works for this special use of `observer.error` and none of the other ones.

`console.error` seems most likely to be lost out of the different alternatives. It's not much better than silently failing.

My rationale for calling `observer.error` is this: There are two types of synchronous errors that might happen:
1. Coding errors that the caller can't be expected to handle. For example, "you passed me both an `observer` *and* an `onError` function." This is a clear coding error, so it's fine to throw a synchronous exception.
2. All other errors should be passed to the observer.

I don't think subscribing after a stream has completed is a coding error, since it's not unrecoverable. It's possible that the caller has done it on purpose to make sure that the observable is consumed. I'm also not concerned about `observer.error` being undefined. If the caller hasn't passed one, then it's clear that they don't care about handling errors.

Reply:
Thanks for the explanation, I'll implement it as you suggest. I'm a bit weary that users may never know they're subscribing twice but as you said, it's a recoverable error, and not a coding error.